kong/tools/queue.lua

```diff
@@ -0,0 +1,385 @@
+--       batch_max_size = 10,    -- maximum number of entries in one batch (default 1)
+--       max_delay = 1,          -- maximum number of seconds after first entry before a batch is sent
+--       capacity = 10,          -- maximum number of entries on the queue (default 10000)
+--       string_capacity = 100,  -- maximum number of bytes on the queue (default nil)
+--       max_retry_time = 60,    -- maximum number of seconds before a failed batch is dropped
+--       max_retry_delay = 60,   -- maximum delay between send attempts
+--     }
+--
+--   Queue.enqueue(queue_conf, handler, handler_conf, "Some value")
+--   Queue.enqueue(queue_conf, handler, handler_conf, "Another value")
+--
+-- Given the example above,
+--
+-- * If the two `enqueue()` invocations are done within `max_delay` seconds, they will be passed to the
+--   handler function together in one batch.  The maximum number of entries in one batch is defined
+--   by the `batch_max_size` parameter.
+-- * The `capacity` parameter defines how many entries can be waiting on the queue for transmission.  If
+--   it is exceeded, the oldest entries on the queue will be discarded when new entries are queued.  Error
+--   messages describing how many entries were lost will be logged.
+-- * The `string_capacity` parameter, if set, indicates that no more than that number of bytes can be
+--   waiting on a queue for transmission.  If it is set, only string entries must be queued and an error
+--   message will be logged if an attempt is made to queue a non-string entry.
+-- * When the `handler` function does not return a true value for a batch, it is retried for up to
+--   `max_retry_time` seconds before the batch is deleted and an error is logged.  Retries are organized
+--   by the queue library using an exponential back-off algorithm with the maximum time between retries
+--   set to `max_retry_delay` seconds.
+
+local semaphore = require "ngx.semaphore"
+
+local DEBUG = ngx.DEBUG
+local INFO = ngx.INFO
+local WARN = ngx.WARN
+local ERR = ngx.ERR
+
+
+local function now()
+  ngx.update_time()
+  return ngx.now()
+end
+
+
+local Queue = {
+  POLL_TIME = 1,
+  MAX_IDLE_TIME = 60,
+}
+
+local Queue_mt = {
+  __index = Queue
+}
+
+
+local queues = {}
+
+
+function Queue.exists(name)
+  return queues[name] and true or false
+end
+
+-------------------------------------------------------------------------------
+-- Initialize a queue with background retryable processing
+-- @param process function, invoked to process every payload generated
+-- @param opts table, requires `name`, optionally includes `retry_count`, `max_delay` and `batch_max_size`
+-- @return table: a Queue object.
+local function get_queue(queue_conf, handler, handler_conf)
+
+  assert(type(queue_conf.name) == "string",
+    "missing name in queue configuration")
+
+  local name = queue_conf.name
+
+  local queue = queues[name]
+  if queue then
+    queue:log(DEBUG, "queue exists")
+    queue.handler_conf = handler_conf
```

Comment:
We always use the last configuration that was given to the queue.  One possible consequence is that during reconfigurations, we may see some oscillation between the old and the new configuration until the new configuration is completely rolled out.  This will happen only when queues are shared.  In the general case, we always use the latest configuration even for entries that have been queued with an earlier configuration.